core/rawdb/freezer_meta.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>
+
+package rawdb
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+	"os"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+const freezerVersion = 1 // The initial version tag of freezer table metadata
+
+// freezerTableMeta wraps all the metadata of the freezer table.
+type freezerTableMeta struct {
+	// version is the versioning descriptor of the freezer table.
+	version uint16
+
+	// VirtualTail indicates how many items have been marked as deleted.
+	// Its value is equal to the number of items removed from the table
+	// plus the number of items hidden in the table, so it should never
+	// be lower than the "actual tail".
+	VirtualTail uint64
+}
+
+// newMetadata initializes the metadata object with the given virtual tail.
+func newMetadata(tail uint64) *freezerTableMeta {
+	return &freezerTableMeta{
+		version:     freezerVersion,
+		VirtualTail: tail,
+	}
+}
+
+// readMetadata reads the metadata of the freezer table from the
+// given metadata file.
+func readMetadata(file *os.File) (*freezerTableMeta, error) {
```

Comment:
If you make `Version` public, then you could just read the whole thing as `rlp`. Not sure why you store it as you do currently - if I understand corectly: 
```
<uint16><rlp( [ uint64 } ) > 
```
Instead of
```
rlp ( [ uint16, uint16 ] ) 
```
If we later decide to extend the meta, we can just add more rlp-items. If we want to read the first `uint16` before continuing, we just need to use the `rlp.List` interface to access the first item in the list.

Reply:
Yes, right now it's `<uint16><rlp( [ uint64 } ) > ` which I use the first two bytes to route the decoding rules.

`rlp ( [ uint16, uint16 ] ) ` also works for me, we can hold the assumption the first field is always this version tag. I can try with this approach to see if the code can get cleaner.

Reply:
Looks like it has more code

```
        _, err := file.Seek(0, io.SeekStart)
	if err != nil {
		return nil, err
	}
	s := rlp.NewStream(file, 0)
	if _, err := s.List(); err != nil {
		return nil, err
	}
	val, err := s.Uint()
	if err != nil {
		return nil, err
	}
	switch val {
	case freezerVersion:
		// decode version 1
		var meta freezerTableMeta
		file.Seek(0, io.SeekStart)
		rlp.Decode(file, &meta)
		// ...
	}
```

I would still prefer the `<uint16><rlp( [ uint64 } ) >` struct. We use the similar struct for `receipt` with a prefix byte for routing. What do you think @holiman

Reply:
Well, it could also be done like this: 
```
func readMetadata(file *os.File) (*freezerTableMeta, error) {
	if _, err := file.Seek(0, io.SeekStart); err != nil {
		return nil, err
	}
	var v freezerTableMeta
	if err := rlp.Decode(file, &v); err != nil {
		return nil, errors.New("incompatible version")
	}
	if v.Version != freezerVersion {
		return nil, errors.New("incompatible version")
	}
	return v, nil
}
```
I don't feel super strongly about it, just seems to make sense to have the entire thing in one encoding-type, instead of mashing two types like that. I may be convinced otherwise.

Reply:
Applied